data/simpy.gram

```diff
@@ -52,153 +52,153 @@ augassign[AugOperator*]:
+        _Py_ExceptHandler(e, NULL, b, EXTRA) }
     | a='except' ':' b=block {
-        _Py_ExceptHandler(NULL, NULL, b, EXTRA(a, token_type, seq_get_tail(NULL, b), stmt_type)) }
+        _Py_ExceptHandler(NULL, NULL, b, EXTRA) }
 finally_block[asdl_seq*]: 'finally' ':' a=block { a }
 
 return_stmt[stmt_ty]:
-    | a='return' b=expressions { _Py_Return(b, EXTRA(a, token_type, b, expr_type)) }
-    | a='return' { _Py_Return(NULL, EXTRA(a, token_type, a, token_type)) }
+    | a='return' b=expressions { _Py_Return(b, EXTRA) }
+    | a='return' { _Py_Return(NULL, EXTRA) }
 
 raise_stmt[stmt_ty]:
-    | a='raise' b=expression 'from' c=expression { _Py_Raise(b, c, EXTRA(a, token_type, c, expr_type)) }
-    | a='raise' b=expression { _Py_Raise(b, NULL, EXTRA(a, token_type, b, expr_type)) }
-    | a='raise' { _Py_Raise(NULL, NULL, EXTRA(a, token_type, a, token_type)) }
+    | a='raise' b=expression 'from' c=expression { _Py_Raise(b, c, EXTRA) }
+    | a='raise' b=expression { _Py_Raise(b, NULL, EXTRA) }
+    | a='raise' { _Py_Raise(NULL, NULL, EXTRA) }
 
 function_def[stmt_ty]:
-    | d=[decorators] s=ASYNC 'def' n=NAME '(' params=parameters ')' a=['->' z=annotation { z }] ':' b=block {
-        _Py_AsyncFunctionDef(((expr_ty) n)->v.Name.id, params, b, d, a, NULL,
-                             EXTRA(s, token_type, seq_get_tail(NULL, b), stmt_type)) }
-    | d=[decorators] s=ASYNC 'def' n=NAME '(' ')' a=['->' z=annotation { z }] ':' b=block {
-        _Py_AsyncFunctionDef(((expr_ty) n)->v.Name.id, empty_arguments(p), b, d, a, NULL,
-                             EXTRA(s, token_type, seq_get_tail(NULL, b), stmt_type)) }
-    | d=[decorators] s='def' n=NAME '(' params=parameters ')' a=['->' z=annotation { z }] ':' b=block {
-        _Py_FunctionDef(((expr_ty) n)->v.Name.id, params, b, d, a, NULL,
-                        EXTRA(s, token_type, seq_get_tail(NULL, b), stmt_type)) }
-    | d=[decorators] s='def' n=NAME '(' ')' a=['->' z=annotation { z }] ':' b=block {
-        _Py_FunctionDef(((expr_ty) n)->v.Name.id, empty_arguments(p), b, d, a, NULL,
-                        EXTRA(s, token_type, seq_get_tail(NULL, b), stmt_type)) }
+    | d=decorators f=function_def_raw { function_def_decorators(p, d, f) }
+    | function_def_raw
+
+function_def_raw[stmt_ty]:
+    | s=ASYNC 'def' n=NAME '(' params=parameters ')' a=['->' z=annotation { z }] ':' b=block {
+        _Py_AsyncFunctionDef(((expr_ty) n)->v.Name.id, params, b, NULL, a, NULL, EXTRA) }
+    | s=ASYNC 'def' n=NAME '(' ')' a=['->' z=annotation { z }] ':' b=block {
+        _Py_AsyncFunctionDef(((expr_ty) n)->v.Name.id, empty_arguments(p), b, NULL, a, NULL, EXTRA) }
+    | s='def' n=NAME '(' params=parameters ')' a=['->' z=annotation { z }] ':' b=block {
+        _Py_FunctionDef(((expr_ty) n)->v.Name.id, params, b, NULL, a, NULL, EXTRA) }
+    | s='def' n=NAME '(' ')' a=['->' z=annotation { z }] ':' b=block {
+        _Py_FunctionDef(((expr_ty) n)->v.Name.id, empty_arguments(p), b, NULL, a, NULL, EXTRA) }
```

Comment:
I suspect we can do some collapsing here too. The last two could share an action:
```
    'def' n=NAME '(' params=[parameters] ')' a=['->' z=annotation { z }] ':' b=block {
        _Py_FunctionDef(((expr_ty) n)->v.Name.id, params ? params : empty_arguments(p), b, NULL, a, NULL, EXTRA) }
```